### ISSUE TITLE
Gossip SGD (as a DDP Communication Hook)

### ISSUE THREAD

**Issue by wayi1** [pytorch/pytorch#71330]
## Need for An Replacement of AllReduce Communication Collective

Nowadays powerful industry users (not necessarily in FAANG) like Cruise, Microsoft, and Tesla are moving towards scaling large-scale training to 100+ or even 1K+ GPUs. DDP, as the most well-adopted data parallelism implementation in PyTorch, all-reduces the gradients of all the processes at each iteration by default. Since allreduce is a communication collective, this does not allow us to effectively target a certain individual process. This can cause a few issues in some scenarios:

1.  If some processes are stragglers, allreduce makes it almost impossible to target these individual stragglers and very hard to mitigate this issue. By contrast, parameter-server architecture can often tackle stragglers in a much easier way, such as discarding the late updates and/or [preparing some backup workers](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/45187.pdf).
2. Let's say if we scale out training to 1K relatively low-end GPUs, fault tolerance could be a concern. The current common solution is restarting from the most recent checkpoint and re-initializing the entire large process group, even if only one worker fails. This can be very inefficient if the large process group initialization is slow and the failure is not rare. Therefore, we may need a finer-grained fault-tolerance mechanism at the communication layer.

I believe one potential solution could be Gossip SGD ([example](https://arxiv.org/abs/1803.05880)), which has no need for a communication collective. Note that Gossip SGD is also mentioned in this [composable DDP proposal](https://github.com/pytorch/pytorch/issues/37002) back in 2020.

### Alternatives

An alternate solution is parameter-server architecture for the synchronization. However, this architecture can also have some other disadvantages over AllReduce, and such solution often requires more drastic changes to the overall architecture/config. In comparison, gossip SGD can be implemented as a communication hook and seamlessly replace the vanilla allreduce in PyTorch.

### Additional context

This can be considered as a DDP feature request. To facilitate the triage:

Labels: feature, module:ddp, oncall:distributed.
CC PyTorch Distributed DDP developers: @mrshenli @pritamdamania87 @zhaojuanmao @rohan-varma

**Comment by rohan-varma** [comment]
@samdow When adding `oncall: distributed` as a label can we avoid adding "triaged" label as well? This ensures it lands up in the distributed team's queue for the oncall to take a look at. Thank you!

**Comment by rohan-varma** [comment]
@wayi1  Thanks for the proposal! Do you happen to have some concrete examples/ideas as to how GossipSGD might improve straggler issue and fault tolerance (avoiding an entire job restart)?

**Comment by wayi1** [comment]
> @wayi1 Thanks for the proposal! Do you happen to have some concrete examples/ideas as to how GossipSGD might improve straggler issue and fault tolerance (avoiding an entire job restart)?

No concrete example yet. I think Gossip SGD can serve as a foundation towards the goal of improving straggler or fault tolerance, and it will still need some additional mechanisms to achieve the goal (e.g., when timeout on a p2p communication, switch to a different neighbor).
